### DISCUSSION TITLE
TMS with TileLayer?

### DISCUSSION THREAD

**Discussion by jfuehner** [visgl/deck.gl#5607] [Q&A]
I am trying to use the [TileLayer](https://deck.gl/docs/api-reference/geo-layers/tile-layer) with a TMS _(MapProxy)_ service. With the TileLayer by default supporting _slippy map tilenames_/(`zxy`) my initial thought was I could simply flip the `y` coordinate (`Math.pow(2, z)-y-1`) to be TMS compliant via the `getTileData`.

Flipping the `y` coordinate within the `getTileData` function appeared to put the tiles in the correct order. However, no matter what my viewState is the map always loads tiles around South America...

For example, if my viewState is...

```
{
   latitude: 37.77,
   longitude: 122.41,
   zoom: 0
}
```

I get tiles loaded looking like the below screen capture. The rest of the world tiles do not load....

![image](https://user-images.githubusercontent.com/1843941/112144122-2a4a9280-8baf-11eb-96e0-a54947d41ab6.png)

I also tried explicitly specifying the TileLayer's `extent` property to the bounds of the TMS service...

`[-20037508.342789244, -20037508.342789244, 20037508.342789244, 20037508.342789244]`

The TMS/MapProxy service provides a test page which uses OpenLayers and is rendering the tiles correctly. I would mockup a Codepen but I can't seem to find a freely available TMS server on the open Internet. 

Appreciate any help and insights on how I can correctly render using deck...

**Comment by kylebarron** [comment]
You can either use an `indexing: tms` prop or use `{-y}` in the data URL field, though it appears that isn't documented on the TileLayer docs page. See https://github.com/visgl/deck.gl/pull/4958

**Reply by jfuehner** [reply]
@kylebarron thanks! Does `indexing: tms` automatically flip the `z` and `y`?

**Reply by kylebarron** [reply]
Oh sorry I meant to write `{-y}` (edited above), the `z` is the same in both; just the `y` is flipped for TMS. And yes setting that prop automatically flips the `y`. A PR to add to the documentation would be great if you're interested.

**Reply by jfuehner** [reply]
@kylebarron was I flipping the `y` too late in the `getTileData`? Was that most likely why the tiles were not displaying appropriately?

**Reply by kylebarron** [reply]
You can look at the referenced PR to see how the flipping is done internally.

**Reply by jfuehner** [reply]
@kylebarron was the `indexing` property removed? Looking through your PR (#4958) it looks like it was renamed at one point to indexingScheme and then indexingScheme was [removed](https://github.com/visgl/deck.gl/pull/4958/commits/5ebb15747fdbda0d18973af3511433a8331ca222).

I also looked at the [tile-layer](https://github.com/visgl/deck.gl/blob/master/modules/geo-layers/src/tile-layer/tile-layer.js) and [tileset-2d](https://github.com/visgl/deck.gl/blob/master/modules/geo-layers/src/tile-layer/tileset-2d.js) files in master and do not see any reference to `indexing` or `indexingScheme` properties.

I also tried the prop with this [Codepen](https://codepen.io/jfuehner/pen/MWJWoby). Adding or removing the property does not seem to impact the tiles...

**Reply by kylebarron** [reply]
I think `indexing` might've been removed before merge. Just use `{-y}` in the data url

**Reply by jfuehner** [reply]
@kylebarron unfortunately for my use case I am not using the `data` property. I am overriding the `getTileData` to call a web socket connection with the `xyz`. 

Unless I can use both the `data` prop and the `getTileData` function?

**Reply by kylebarron** [reply]
The `url` prop is resolved from `data` https://deck.gl/docs/api-reference/geo-layers/tile-layer#gettiledata

**Reply by jfuehner** [reply]
So, I can make using the `data` prop in conjunction with the `getTileData` work. However, it's a bit weird.

If I provide a `data` prop  `/{z}/{x}/{-y}` for example I see in the `getTileData` function `tile` object providing back the resolved `url` prop with the `y` flipped. However, the same `tile` object's `y` prop is the non-flipped value. I would expect the `tile` objects `y` prop to be the flipped value to be consistent with what was requested/provided in the `data` prop?

For my use-case though, the use of the `data` props seems a bit redundant. With my current implementation I don't actually supply a `data` prop to the `TileLayer`. I only provide a handler function to the `getTileData` which calls an already open web socket passing the `tile` object directly to the socket.

With having to use the `data` prop in conjunction with `getTileData` and the `y` values not being consistent I will have to parse the `url` prop out to get the correct `y` value...

**Reply by kylebarron** [reply]
> If I provide a `data` prop `/{z}/{x}/{-y}` for example I see in the `getTileData` function `tile` object providing back the resolved `url` prop with the `y` flipped. However, the same `tile` object's `y` prop is the non-flipped value. I would expect the `tile` objects `y` prop to be the flipped value to be consistent with what was requested/provided in the `data` prop?

I think I raised this before. Might be worth reading through the discussion in the previous PR and the issue that created it https://github.com/visgl/deck.gl/pull/4958#discussion_r496112291 https://github.com/visgl/deck.gl/issues/4757#issuecomment-655717387
